clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,141 @@
+        //--- first_kernel
+        { kernel_param_kind_t::kind_std_layout, 4, 0 },
+        { kernel_param_kind_t::kind_std_layout, 8, 4 },
+        { kernel_param_kind_t::kind_accessor, 4062, 12 },
+        { kernel_param_kind_t::kind_sampler, 8, 24 },
+
+        { kernel_param_kind_t::kind_invalid, -987654321, -987654321 },
+      };
+    } // namespace detail
+  } // namespace _V1
+} // namespace sycl
+```
+
+For each kernel, a mangled name associated with it, is stored in the kernel_names array.
+
+And for each kernel, each of its parameter is described in the kernel_signatures array.   The description consists of the following parts:
+
+* an encoding of the type that is captured
+* its size and its offset
+
+In the above example, the first two parameters are standard layout types (the int 'i' and the struct 'test_s'), the latter two though standard layout types are types which need special handling - accessor and sampler respectively.   The last entry acts as a terminating entry.
+
+This allows the run-time library to recreate the function object on the device.
+
+The full set of encoding is:
+
+* kind_accessor  for sycl accessor types
+* kind_std_layout for standard layout types
+* kind_sampler for sycl sampler types
+* kind_pointer for pointer types
+* kind_specialization_constants_buffer for initializing specialization constants
+* kind_stream for sycl stream types and
+* kind_invalid used to indicate the terminal value in the signatures array.
+
+
+# Integration Footer
+An integration footer is a source file produced during device compilation that is intended to be post-included during host compilation of the same source file.
+
+The SYCL 2020 Specification defines a specialization constant as a constant variable where the value is not known until compilation of the SYCL kernel function.
+
+Consider the following example:
+```
+  constexpr specialiation_id<int> int_const;
+  class Wrapper {
+  public:
+    static constexpr specialization_id<float> float_const;
+  };
+```
+
+Also note that the SYCL header files declare, but not define, a special function to obtain the numeric id of a specialization constant.
+```
+     namespace detail {
+         template<auto &SpecConstName>
+         inline const char *get_spec_constant_symbolic_ID();
+     }
+```
+
+In order to accommodate both implementations that have native support for this as well as those that do not, we propose the following integration footer mechanism.
+
+Definition of that function template is provided in the form of an integration footer file:
+```
+   namespace detail {
+     template<>
+     inline const char *get_spec_constant_symbolic_ID<int_const>() {
+       return "unique_name_for_int_const";
+     }
+     template<>
+     inline const char *get_spec_constant_symbolic_ID<Wrapper::float_const>() {
+       return "unique_name_for_Wrapper_float_const";
+     }
+   }
+```
+
+This footer file is appended at the end of the translation unit for the host compilation.
```

Comment:
For the integration header, there is a pretty clear story for how to prepend the header. Most compilers support something like gcc and clang's `-include`. For the same is not true for an integration footer. I think the RFC should propose how this would be supported for third party host compilers.

Reply:
I sent out my proposed text for the open question on Friday via email.  There is a section titled "Mechanism for integrating the host and device code", which raises the possibility of using a different mechanism from the integration header / footer.

I already note some deficiencies with the integration header approach.  If you know of other deficiencies, it would make sense to note them in that same section.

I think we were planning to add this text to the main RFC document.  Therefore, I'll let you, @premanandrao, incorporate the text into your PR and augment it with information about other deficiencies with the integration header approach.